File: distancesEuclidean.lua
-- distancesEuclidean.lua

require 'makeVp'

-- determine Euclidean distances to a query point
--
-- ARGS:
--
-- inputs: 2D tensor of size n x d
--
-- query: 1D Tensor of size d
--
-- RETURN 1D Tensor of size n
-- with Euclidean distance for query to each input row
function distancesEuclidean(inputs, query)
   assert(inputs:nDimension() == 2)
   local n = inputs:size(1)
   local d = inputs:size(2)
   assert(query:nDimension() == 1)
   assert(query:size(1) == d)

   verbose = 0
   local vp = makeVp(verbose)

   vp(1, 'inputs', inputs)
   vp(1, 'query', query)

   -- create synthetic query2D is size n x d
   local query2D = torch.Tensor(query:storage(),
                                1,  -- storageOffset
                                n, 0, -- sz1, st1
                                d, 1) -- sz2, st2
   -- test that we did this correctly
   vp(2, 'query2D', query2D)
   assert(query2D:nDimension() == 2)
   assert(query2D:size(1) == n)
   assert(query2D:size(2) == d)
   assert(query[1] == query2D[3][1])
   if d > 1 then
      assert(query[2] == query2D[1][2])
   end
   if d > 2 then
      assert(query[3] == query2D[2][3])
   end

   local diffs = inputs - query2D
   vp(2, 'diffs', diffs)
   local squares = torch.cmul(diffs, diffs)
  
   local sumSquares = torch.Tensor(n):zero()
   for i = 1, d do
      sumSquares = torch.add(sumSquares, squares:select(2, i))
   end

   local result = sumSquares:sqrt()
   vp(1, 'result', result)
   return result
end

File: DistancesEuclidean.R
# DistancesEuclidean.R

DistancesEuclidean <- function(df, query, featureNames) {
  # Determine Euclidean distances from all observations in a data frame
  # to a query observation.
  #
  # ARGS
  #
  # df: data frame containing the observations
  #
  # query: a data frame with one observations
  #
  # featureNames: vector or list of character, the names of the features in df
  # and query to use
  #
  # Value: numeric vector containing the Euclidean distance from each
  # observation in df to the query point

  if (nrow(query) != 1)
    stop("query must have one row")
  if (class(featureNames) != "character")
    stop("featureNames class must be character")
  sum <- 0
  for (featureName in featureNames) {
  	distances <- df[[featureName]] - query[[featureName]]
  	sum <- sum + distances * distances
  }
  result <- sqrt(sum)
  if (length(result) == 0) {
    print("DistancesEuclidean: result has zero length")
    browser()
  }
  result
}

DistancesEuclidean.Test <- function() {
  # Unit test of DistancesEuclidean
  df <- data.frame(x=c(1,2),
                   y=c(3,4))
  query <- data.frame(x=4,
                      y=5)
  featureNames <- c("x", "y")
  distances <- DistancesEuclidean(df, query, featureNames)
  # distances[1] = sqrt((1-4)^2 + (3-5)^2) = sqrt(9 + 4)
  # distances[2] = sqrt((2-4)^2 + (4-5)^2) = sqrt(4 + 1)
  tolerance <- 1e-6
  Near <- function(expected, actual) {
    if (abs(expected - actual) > tolerance)
      stop(sprintf("unit test error; expected=%f actual=%f",
                   expected, actual))
  }
  Near(sqrt(13), distances[1])
  Near(sqrt(5), distances[2])
}

DistancesEuclidean.Test()
